File: tests/extraction_test.py
import tempfile
import unittest
from pathlib import Path
import pymupdf
from backend_test import bridge


class ExtractionTests(unittest.TestCase):
    def fixture(self, root, kind):
        doc = pymupdf.open()
        page = doc.new_page()
        text = "Representative born-digital evidence from the preserved document. " * 20
        if kind == "columns":
            page.insert_textbox(pymupdf.Rect(40, 40, 250, 750), text)
            page.insert_textbox(pymupdf.Rect(310, 40, 550, 750), text)
        elif kind == "tables":
            page.insert_text((40, 60), "Year     Mean     Uncertainty\n2020     4.1      0.2\n2021     5.2      0.3\nTable layout matters for the evidence.")
            page.draw_line((40, 80), (300, 80))
        elif kind == "equations":
            page.insert_text((40, 60), "Conservation model: E = m c^2. Superscripts and symbols require visual review.")
        elif kind != "scan":
            page.insert_textbox(pymupdf.Rect(40, 40, 550, 750), text)
        if kind in ("scan", "missing"):
            # Drawn content with no extractable text simulates image-only pages.
            blank = page if kind == "scan" else doc.new_page()
            blank.draw_rect(pymupdf.Rect(40, 40, 500, 700), fill=(0.8, 0.8, 0.8))
        body = "\n\n---\n\n".join(page.get_text("text") for page in doc if page.get_text("text").strip()).strip()
        path = root / "research" / "raw" / "fixture.pdf"
        path.parent.mkdir(parents=True)
        doc.save(path)
        doc.close()
        return {"source": "https://example.org/fixture.pdf", "raw_file": "raw/fixture.pdf", "oa": {"url": "https://repo.example/accepted.pdf", "version": "acceptedVersion"}}, body

    def test_text_pages_have_pinned_assets_and_real_page_numbers(self):
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            meta, body = self.fixture(root, "text")
            result, spans = bridge.extraction_diagnostics(root, meta, body, 0, len(body))
            self.assertEqual(result["status"], "text-extracted")
            self.assertEqual(result["pages"], 1)
            self.assertEqual(result["version"], "accepted")
            self.assertEqual(result["actualUrl"], "https://repo.example/accepted.pdf")
            self.assertEqual(len(result["rawHash"]), 64)
            self.assertEqual(spans[0]["page"], 1)
            self.assertTrue(body[spans[0]["start"]:spans[0]["end"]])

    def test_scans_and_missing_pages_never_pass_as_complete(self):
        for kind in ("scan", "missing"):
            with self.subTest(kind=kind), tempfile.TemporaryDirectory() as tmp:
                root = Path(tmp)
                meta, body = self.fixture(root, kind)
                result, spans = bridge.extraction_diagnostics(root, meta, body, 0, len(body))
                self.assertEqual(result["status"], "incomplete")
                self.assertEqual(result["missingPages"], [1] if kind == "scan" else [2])
                self.assertTrue(any("OCR" in warning for warning in result["warnings"]))

    def test_columns_tables_and_equations_do_not_imply_visual_comprehension(self):
        for kind, warning in (("columns", "layout"), ("tables", "tables"), ("equations", "equations")):
            with self.subTest(kind=kind), tempfile.TemporaryDirectory() as tmp:
                root = Path(tmp)
                meta, body = self.fixture(root, kind)
                result, _ = bridge.extraction_diagnostics(root, meta, body, 0, len(body))
                self.assertEqual(result["status"], "text-extracted")
                self.assertIn(warning + "-unverified", result["warnings"])

    def test_missing_corrupt_or_escaped_assets_fail_closed(self):
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)
            meta, body = self.fixture(root, "text")
            with self.assertRaisesRegex(ValueError, "Unsafe"):
                bridge.extraction_diagnostics(root, {**meta, "raw_file": "../private.pdf"}, body, 0, len(body))
            path = root / "research" / "raw" / "fixture.pdf"
            path.write_bytes(b"%PDF-invalid")
            self.assertEqual(bridge.extraction_diagnostics(root, meta, body, 0, len(body))[0]["status"], "unavailable")
            path.unlink()
            self.assertEqual(bridge.extraction_diagnostics(root, meta, body, 0, len(body))[0]["status"], "unknown")

File: tests/extraction.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";
import { mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";
import { PDFDocument, StandardFonts } from "pdf-lib";
import { extractPdf, extractSource, hashBytes } from "../src/extraction.ts";
import { adequateExtraction } from "../src/evidence.ts";
import { SourceStore } from "../src/source-store.ts";
import { privateDirectory } from "../src/paths.ts";

export async function pdfFixture(kind: "text" | "scan" | "missing" | "columns" | "tables" | "equations" = "text") {
  const doc = await PDFDocument.create(); const font = await doc.embedFont(StandardFonts.Helvetica);
  const page = doc.addPage();
  const text = kind === "tables" ? "Year    Mean    Uncertainty\n2020    4.1     0.2\n2021    5.2     0.3\nTable layout requires review."
    : kind === "equations" ? "Conservation model: E = m c^2. Superscripts and symbols require visual review."
    : "Representative born-digital evidence from the preserved document. ";
  if (kind !== "scan") page.drawText(text, { x: 40, y: 700, size: 10, font, maxWidth: kind === "columns" ? 200 : 500 });
  if (kind === "columns") page.drawText(text, { x: 300, y: 700, size: 10, font, maxWidth: 200 });
  if (kind === "scan") page.drawRectangle({ x: 40, y: 40, width: 400, height: 600 });
  if (kind === "missing") doc.addPage().drawRectangle({ x: 40, y: 40, width: 400, height: 600 });
  return doc.save();
}

test("Readability selects static article text without executing scripts or retaining navigation", async () => {
  const body = "Measured source evidence explains the outcome in detail. ".repeat(60);
  const html = `<html><head><title>Research article</title><meta name="citation_doi" content="10.1234/paper"></head><body><nav>PRIVATE NAVIGATION</nav><article><h1>Research article</h1><p>${body}</p><p>${body}</p></article><script>throw new Error('EXECUTED')</script><footer>FOOTER NOISE</footer></body></html>`;
  const result = await extractSource({ url: "https://example.org/article", bytes: Buffer.from(html), contentType: "text/html" });
  assert.match(result.body, /Measured source evidence/); assert.ok(!/PRIVATE NAVIGATION|EXECUTED|FOOTER NOISE/.test(result.body));
  assert.equal(result.doi, "10.1234/paper"); assert.equal(result.extraction.status, "text-extracted");
  for (const requirement of ["layout", "tables", "figures", "equations"]) assert.equal(adequateExtraction(result.extraction, [requirement]), false);
});

test("blocked, sparse, binary and malformed text are not accepted as readable articles", async () => {
  for (const response of [
    { contentType: "text/html", bytes: Buffer.from("<html><head><title>Verify you are human</title></head><body><article>" + "Complete the captcha. ".repeat(50) + "</article></body></html>") },
    { contentType: "text/plain", bytes: Buffer.from("tiny") },
    { contentType: "application/octet-stream", bytes: Buffer.from([0, 1, 2, 3]) },
    { contentType: "text/plain; charset=utf-8", bytes: Buffer.from([0xff, 0xff]) },
  ]) await assert.rejects(extractSource({ url: "https://example.org/source", ...response }));
});

test("PDF.js pins preserved assets and physical page spans without claiming visual understanding", async () => {
  for (const kind of ["text", "columns", "tables", "equations"] as const) {
    const bytes = await pdfFixture(kind); const result = await extractPdf(bytes, "https://example.org/source.pdf");
    assert.equal(result.extraction.status, "text-extracted", JSON.stringify(result)); assert.equal(result.extraction.pages, 1);
    assert.equal(result.extraction.rawHash, hashBytes(bytes)); assert.equal(result.pageSpans[0].page, 1);
    assert.ok(result.body.slice(result.pageSpans[0].start, result.pageSpans[0].end).length >= 40);
    assert.equal(adequateExtraction(result.extraction, [kind === "text" || kind === "columns" ? "layout" : kind]), false);
  }
});

test("scanned, sparse, corrupt and cancelled PDFs never count as complete reads", async () => {
  for (const kind of ["scan", "missing"] as const) {
    const result = await extractPdf(await pdfFixture(kind), "https://example.org/source.pdf");
    assert.equal(result.extraction.status, "incomplete"); assert.deepEqual(result.extraction.missingPages, [kind === "scan" ? 1 : 2]);
    assert.equal(adequateExtraction(result.extraction), false); assert.ok(result.extraction.warnings.some(warning => warning.includes("OCR")));
  }
  const corrupt = await extractPdf(Buffer.from("%PDF-invalid"), "https://example.org/source.pdf");
  assert.equal(corrupt.extraction.status, "unavailable");
  const abort = new AbortController(); abort.abort();
  await assert.rejects(extractPdf(await pdfFixture(), "https://example.org/source.pdf", abort.signal));
  const midParse = new AbortController();
  const pending = extractPdf(await pdfFixture(), "https://example.org/source.pdf", midParse.signal);
  setTimeout(() => midParse.abort(), 5);
  await assert.rejects(pending);
  const oversized = await PDFDocument.create(); for (let i = 0; i < 301; i++) oversized.addPage();
  assert.equal((await extractPdf(await oversized.save(), "https://example.org/long.pdf")).extraction.status, "unavailable");
});

test("legacy PDF diagnostics never rewrite preserved bodies; incompatible text and missing assets fail closed", async () => {
  const root = mkdtempSync(join(tmpdir(), "hpr-legacy-pdf-"));
  try {
    privateDirectory(join(root, "research", "notes")); privateDirectory(join(root, "research", "raw"));
    const bytes = await pdfFixture(); const parsed = await extractPdf(bytes, "https://repo.example/source.pdf");
    const asset = join(root, "research", "raw", "source.pdf"); writeFileSync(asset, bytes);
    const file = join(root, "research", "notes", "legacy-pdf.md");
    const header = "---\nid: legacy-pdf\ntitle: Legacy PDF\nsource: https://example.org/source.pdf\nraw_file: raw/source.pdf\noa_url: https://repo.example/source.pdf\noa_version: acceptedVersion\n---\n\n";
    writeFileSync(file, header + parsed.body);
    const store = new SourceStore(root); const before = readFileSync(file);
    const read = await store.page("legacy-pdf");
    assert.equal(read.hash, hashBytes(parsed.body)); assert.equal(read.extraction.status, "text-extracted");
    assert.equal(read.extraction.version, "accepted"); assert.equal(read.extraction.actualUrl, "https://repo.example/source.pdf");
    assert.equal(read.pageSpans[0].page, 1); assert.deepEqual(readFileSync(file), before);
    writeFileSync(file, header + "Changed text mapping");
    assert.equal((await store.page("legacy-pdf")).extraction.status, "incomplete");
    rmSync(asset);
    assert.equal((await store.page("legacy-pdf")).extraction.status, "unknown");
  } finally { rmSync(root, { recursive: true, force: true }); }
});
